File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/hub/FlutterHubElement.kt
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package com.amazonaws.amplify.amplify_datastore.types.hub

import com.amazonaws.amplify.amplify_datastore.types.model.FlutterSerializedModel
import com.amplifyframework.core.model.Model
import com.amplifyframework.core.model.SerializedModel
import com.amplifyframework.datastore.syncengine.OutboxMutationEvent

data class FlutterHubElement(
    private val modelWithMetadata: OutboxMutationEvent.OutboxMutationEventElement<out Model>
) {
    fun toValueMap(): Map<String, Any> {
        val serializedModel: MutableMap<String, Any> = FlutterSerializedModel(
            this.modelWithMetadata.model as SerializedModel
        ).toMap().toMutableMap()
        return mapOf(
            "syncMetadata" to FlutterSyncMetaData(this.modelWithMetadata).toValueMap(),
            "model" to serializedModel
        )
    }
}

File: packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import Foundation
import Flutter
import UIKit
import Combine

public struct FlutterHubElement {
    var model: [String: Any]
    var version: Int?
    var lastChangedAt: Int64?
    var deleted: Bool

    init(
        hubElement: OutboxMutationEvent.OutboxMutationEventElement,
        modelSchemaRegistry: FlutterSchemaRegistry,
        customTypeSchemaRegistry: FlutterSchemaRegistry,
        modelName: String
    ) throws {
        guard let model = hubElement.model as? FlutterSerializedModel else {
            throw FlutterDataStoreError.hubEventCast
        }
        self.model = try model.toMap(
            modelSchemaRegistry: modelSchemaRegistry,
            customTypeSchemaRegistry: customTypeSchemaRegistry,
            modelName: modelName
        )
        self.version = hubElement.version
        self.lastChangedAt = hubElement.lastChangedAt
        self.deleted = hubElement.deleted ?? false
    }

    init(
        hubElement: MutationEvent,
        modelSchemaRegistry: FlutterSchemaRegistry,
        customTypeSchemaRegistry: FlutterSchemaRegistry,
        modelName: String
    ) throws {
        do {
            guard let decodedModel = (try ModelRegistry.decode(modelName: modelName, from: hubElement.json) as? FlutterSerializedModel) else {
                throw FlutterDataStoreError.hubEventCast
            }
            self.model = try decodedModel.toMap(
                modelSchemaRegistry: modelSchemaRegistry,
                customTypeSchemaRegistry: customTypeSchemaRegistry,
                modelName: modelName
            )
            self.version = hubElement.version
            self.deleted = self.model["_deleted"] as? Bool ?? false
            if let value = self.model["_lastChangedAt"] as? Double {
                self.lastChangedAt = Int64(value)
            } else if let value = self.model["_lastChangedAt"] as? String {
                self.lastChangedAt = Int64(value)
            } else if let value = self.model["_lastChangedAt"] as? Int {
                self.lastChangedAt = Int64(value)
            }
        } catch {
            throw FlutterDataStoreError.hubEventCast
        }
    }

    func toValueMap() -> [String: Any] {
        return [
            "model": model,
            "syncMetadata": [
                "_version": version as Any,
                "_lastChangedAt": lastChangedAt as Any,
                "_deleted": deleted as Any
            ]
        ]
    }
}
